src/rust/inetstack/protocols/layer4/tcp/established/ctrlblk.rs

```diff
@@ -661,47 +625,27 @@ impl SharedControlBlock {
-                // and update the sender window.
-                self.sender.process_ack(cb, header, now);
-                if header.ack_num == send_next {
-                    // This segment acknowledges everything we've sent so far (i.e. nothing is currently outstanding).
-
-                    // Since we no longer have anything outstanding, we can turn off the retransmit timer.
-                    self.send_retransmit_deadline_time_secs.set(None);
-                } else {
-                    // Update the retransmit timer.  Some of our outstanding data is now acknowledged, but not all.
-                    // TODO: This looks wrong.  We should reset the retransmit timer to match the deadline for the
-                    // oldest still-outstanding data.  The below is overly generous (minor efficiency issue).
-                    let deadline: Instant = now + self.send_rto_calculator.rto();
-                    self.send_retransmit_deadline_time_secs.set(Some(deadline));
-                }
-
-                let nbytes: usize = Into::<u32>::into(header.ack_num - send_unacknowledged) as usize;
-                self.receive_ack_queue_frame_bytes.push(nbytes);
-            } else {
-                // This segment acknowledges data we have yet to send!?  Send an ACK and drop the segment.
-                // TODO: See RFC 5961, this could be a Blind Data Injection Attack.
-                let cause: String = format!("Received segment acknowledging data we have yet to send!");
-                warn!("process_ack(): {}", cause);
-                self.send_ack();
-                return Err(Fail::new(libc::EBADMSG, &cause));
-            }
+        // Check whether this is an ack for data that we have sent.
+        if header.ack_num <= send_next {
+            // Does not matter when we get this since the clock will not move between the beginning of packet
+            // processing and now without a call to advance_clock.
+            let now: Instant = self.get_now();
+            self.sender.process_ack(header, now);
+            let nbytes: usize = Into::<u32>::into(header.ack_num - send_unacknowledged) as usize;
+            self.receive_ack_queue_frame_bytes.push(nbytes)
         } else {
-            // Duplicate ACK (doesn't acknowledge anything new).  We can mostly ignore this, except for fast-retransmit.
-            // TODO: Implement fast-retransmit.  In which case, we'd increment our dup-ack counter here.
-            warn!("process_ack(): received duplicate ack ({:?})", header.ack_num);
+            // This segment acknowledges data we have yet to send!?  Send an ACK and drop the segment.
+            // TODO: See RFC 5961, this could be a Blind Data Injection Attack.
+            let cause: String = format!("Received segment acknowledging data we have yet to send!");
+            warn!("process_ack(): {}", cause);
+            self.send_ack();
```

Comment:
Should this code send an ack if this potentially and attack?

Reply:
Yes, this is what the RFC says that we must do. I thought it was weird too so I actually looked up [RFC 5961](https://datatracker.ietf.org/doc/html/rfc5961)